sound/soc/sof/topology.c

```diff
@@ -1959,14 +1964,32 @@ static int sof_widget_unload(struct snd_soc_component *scomp,
 
 		break;
 	case snd_soc_dapm_pga:
-
-		/* get volume kcontrol */
-		kc = &widget->kcontrol_news[0];
-		sm = (struct soc_mixer_control *)kc->private_value;
-		scontrol = sm->dobj.private;
-
-		/* free volume table */
-		kfree(scontrol->volume_table);
+		for (i = 0; i < widget->num_kcontrols; i++) {
+			kc = &widget->kcontrol_news[i];
+			sm = (struct soc_mixer_control *)kc->private_value;
+			scontrol = sm->dobj.private;
+
+			/*
+			 * volume_table is allocated
+			 * only if num_kcontrols = 1
```

Comment:
@bardliao actually the more accurate reason for allocating the volume table is if sm->max > 1

Reply:
@ranj063 But we will not call set_up_volume_table() [if (le32_to_cpu(tw->num_kcontrols) != 1)](https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/sof/topology.c#L1378)

Reply:
@bardliao thats for the pga widget. What about other widgets? there's no reason to have a volume control for other widgets but that doesnt prevent it. What I meant was that if there's a mixer control for a widget, we should set up a volume table only if sm->max > 1.

Reply:
@ranj063 ok I got your point. It's another topic for me. We can do it in the future and it is safe to free NULL. So, let's change the testing condition as it should be.